File: src/cass/share.py
"""`cass share` — ephemeral share links for Claude Code conversations.

Turns a Claude Code session .jsonl into sanitized markdown, uploads to the
cassandra-share service, and prints a clipboard-ready message.

Subcommands:
  cass share create [SESSION]  — create a share, print clipboard text
  cass share list              — list your active shares
  cass share revoke TOKEN      — revoke early
"""
from __future__ import annotations

import json
import os
import re
import subprocess
import sys
from pathlib import Path

import click
import httpx

from cass.auth import ensure_auth


SHARE_URL_DEFAULT = "https://share.cassandrasedge.com"  # admin host (CF Access)


def _share_url() -> str:
    """Base URL for the share service's admin routes.

    Public capability URLs (returned by the server) use a separate
    SHARE_PUBLIC_DOMAIN and are generated server-side — the client doesn't need
    to know about them.
    """
    return os.environ.get("CASS_SHARE_URL") or SHARE_URL_DEFAULT


def _share_headers() -> dict[str, str]:
    """Auth headers for the share service. Requires CF Access login."""
    auth = ensure_auth()
    headers: dict[str, str] = {"Content-Type": "application/json"}
    if auth.get("cf_token"):
        headers["Cf-Access-Jwt-Assertion"] = auth["cf_token"]
    # Dev shortcut: if running against a local share service without CF Access,
    # the service accepts X-Dev-Email.
    if os.environ.get("CASS_DEV_EMAIL"):
        headers["X-Dev-Email"] = os.environ["CASS_DEV_EMAIL"]
    elif auth.get("email"):
        headers["X-Dev-Email"] = auth["email"]
    return headers


# ── Session discovery ─────────────────────────────────────────────────────


def _resolve_session_path(arg: str | None) -> Path:
    """Find the Claude Code session JSONL to share.

    Resolution order:
      1. explicit path argument (if given)
      2. $CLAUDE_SESSION_ID + $CLAUDE_PROJECT_DIR lookup
      3. newest .jsonl in the current-cwd-hashed project dir
    """
    if arg:
        p = Path(arg).expanduser()
        if not p.exists():
            raise click.ClickException(f"No such file: {p}")
        return p

    base = Path.home() / ".claude" / "projects"
    if not base.exists():
        raise click.ClickException(f"{base} missing — Claude Code not set up on this machine")

    session_id = os.environ.get("CLAUDE_SESSION_ID")
    if session_id:
        for project_dir in base.iterdir():
            candidate = project_dir / f"{session_id}.jsonl"
            if candidate.exists():
                return candidate

    # Fallback: newest jsonl under the hashed dir for this cwd.
    cwd_hash_dir = _project_dir_for_cwd(base)
    if cwd_hash_dir and cwd_hash_dir.is_dir():
        candidates = sorted(cwd_hash_dir.glob("*.jsonl"), key=lambda p: p.stat().st_mtime, reverse=True)
        if candidates:
            return candidates[0]

    raise click.ClickException(
        "Could not locate the current session .jsonl. "
        "Pass the path explicitly: cass share create /path/to/session.jsonl"
    )


def _project_dir_for_cwd(base: Path) -> Path | None:
    """Claude Code stores sessions under a directory whose name is the cwd path
    with slashes replaced by hyphens. Try to match the current cwd."""
    cwd = os.getcwd()
    mangled = "-" + cwd.replace("/", "-")
    candidate = base / mangled
    if candidate.exists():
        return candidate
    return None


# ── JSONL → markdown + local sanitizer ────────────────────────────────────


SECRET_PATTERNS = [
    (re.compile(r"sk-[A-Za-z0-9_\-]{20,}"), "<OPENAI_KEY>"),
    (re.compile(r"ghp_[A-Za-z0-9]{30,}"), "<GITHUB_TOKEN>"),
    (re.compile(r"AKIA[0-9A-Z]{16}"), "<AWS_ACCESS_KEY_ID>"),
    (re.compile(r"AIza[0-9A-Za-z_\-]{35}"), "<GOOGLE_API_KEY>"),
    (re.compile(r"sk_live_[A-Za-z0-9]{20,}"), "<STRIPE_KEY>"),
    (re.compile(r"-----BEGIN [A-Z ]+PRIVATE KEY-----[\s\S]+?-----END [A-Z ]+PRIVATE KEY-----"),
     "<PRIVATE_KEY_PEM>"),
    (re.compile(r"eyJ[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+"), "<JWT>"),
]
PATH_PATTERN = re.compile(r"/Users/[a-zA-Z0-9._-]+")
HOSTNAME_PATTERN = re.compile(r"\b(?:\d{1,3}\.){3}\d{1,3}\b")


def _sanitize(text: str) -> str:
    """Local regex sanitizer. Catches well-known credential formats, absolute
    home paths, and private IPs — zero dependencies, always runs.

    For deeper context-aware scanning (JWTs in prose, PEMs across line breaks,
    passwords inside URLs), set CASS_SHARE_DEEP_SCAN=1 to additionally run
    OpenAI Privacy Filter (OPF) via the separate cassandra-scrub endpoint.
    OPF is slower (seconds per 100k tokens on CUDA) but catches things regex
    can't describe. If OPF isn't reachable, we fall back silently to regex-only.
    """
    for pat, repl in SECRET_PATTERNS:
        text = pat.sub(repl, text)
    text = PATH_PATTERN.sub("<HOME>", text)
    text = HOSTNAME_PATTERN.sub("<INTERNAL_IP>", text)

    if os.environ.get("CASS_SHARE_DEEP_SCAN"):
        text = _opf_sanitize(text)
    return text


def _opf_sanitize(text: str) -> str:
    """Deep scan via OPF. No-op if the scrub service isn't configured or fails.

    Contract: POSTs the text to CASS_SCRUB_URL (or a future `cassandra-scrub`
    deployment), which runs OPF server-side and returns the redacted text.
    Keeping the network call optional lets cass ship without a heavy local
    model dependency.
    """
    scrub_url = os.environ.get("CASS_SCRUB_URL")
    if not scrub_url:
        return text
    try:
        with httpx.Client(timeout=60) as client:
            resp = client.post(scrub_url, json={"text": text})
            resp.raise_for_status()
            return resp.json().get("redacted", text)
    except Exception as e:
        click.echo(f"(deep-scan unavailable: {e} — using regex-only)", err=True)
        return text


def _extract_content(record: dict) -> list[tuple[str, str]]:
    """Flatten one record's content into (kind, text) tuples."""
    m = record.get("message")
    if not isinstance(m, dict):
        return []
    role = str(m.get("role", "unknown"))
    content = m.get("content")
    if isinstance(content, str):
        return [(role, content)]
    if not isinstance(content, list):
        return []
    out: list[tuple[str, str]] = []
    for part in content:
        if not isinstance(part, dict):
            continue
        t = part.get("type")
        if t == "text" or (t is None and "text" in part):
            out.append((role, str(part.get("text", ""))))
        elif t == "tool_use":
            name = part.get("name", "tool")
            inp = part.get("input")
            try:
                serial = json.dumps(inp, ensure_ascii=False, indent=2)
            except TypeError:
                serial = str(inp)
            out.append((f"{role}:tool_use:{name}", serial))
        elif t == "tool_result":
            tc = part.get("content")
            if isinstance(tc, str):
                out.append((f"{role}:tool_result", tc))
            elif isinstance(tc, list):
                for y in tc:
                    if isinstance(y, dict) and isinstance(y.get("text"), str):
                        out.append((f"{role}:tool_result", y["text"]))
    return out


def _jsonl_to_markdown(jsonl_path: Path, title: str | None = None) -> str:
    """Render a Claude Code session .jsonl as an LLM-pasteable markdown transcript."""
    lines: list[str] = []
    if title:
        lines.append(f"# {title}")
        lines.append("")

    with jsonl_path.open("r") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                rec = json.loads(line)
            except json.JSONDecodeError:
                continue
            for kind, text in _extract_content(rec):
                if not text.strip():
                    continue
                text = _sanitize(text)
                if kind.endswith(":tool_use") or ":tool_use:" in kind:
                    lines.append(f"### {kind}")
                    lines.append("```json")
                    lines.append(text)
                    lines.append("```")
                elif kind.endswith(":tool_result"):
                    lines.append(f"### {kind}")
                    lines.append("```")
                    lines.append(text[:8000])  # cap giant outputs
                    if len(text) > 8000:
                        lines.append(f"... ({len(text) - 8000} more chars truncated)")
                    lines.append("```")
                else:
                    lines.append(f"## {kind}")
                    lines.append(text)
                lines.append("")
    return "\n".join(lines).strip() + "\n"


# ── TTL parsing ────────────────────────────────────────────────────────────


def _parse_ttl(ttl: str) -> int:
    """Parse '24h', '6h', '7d' into hours."""
    m = re.fullmatch(r"(\d+)([hd])", ttl.strip().lower())
    if not m:
        raise click.BadParameter(f"invalid --ttl {ttl!r}; use forms like 6h, 24h, 7d")
    n, unit = int(m.group(1)), m.group(2)
    return n if unit == "h" else n * 24


# ── Clipboard ──────────────────────────────────────────────────────────────


def _copy(text: str) -> bool:
    """Copy to macOS clipboard via pbcopy. Returns True on success."""
    try:
        p = subprocess.run(["pbcopy"], input=text, text=True, check=True)
        return p.returncode == 0
    except (FileNotFoundError, subprocess.CalledProcessError):
        return False


# ── Commands ───────────────────────────────────────────────────────────────


@click.group()
def share() -> None:
    """Share Claude Code conversations via ephemeral URLs."""


@share.command("create")
@click.argument("session", required=False)
@click.option("--ttl", default="24h", help="Expiry: e.g. 6h, 24h, 7d. Default 24h.")
@click.option("--once", is_flag=True, help="Self-destruct after first fetch.")
@click.option("--title", default=None, help="Optional human title for listing.")
@click.option("--summary", default=None,
              help="2-3 line 'About:' blurb. If omitted, first turn's first 200 chars.")
@click.option("--no-copy", is_flag=True, help="Skip copying to clipboard.")
def create_cmd(session: str | None, ttl: str, once: bool,
               title: str | None, summary: str | None, no_copy: bool) -> None:
    """Upload the current (or specified) session as a share link."""
    jsonl_path = _resolve_session_path(session)
    click.echo(f"Reading {jsonl_path}...", err=True)
    body = _jsonl_to_markdown(jsonl_path, title=title)
    click.echo(f"  → {len(body):,} chars of sanitized markdown", err=True)

    if summary is None:
        # Crude default: first non-empty user line.
        for line in body.splitlines():
            s = line.strip()
            if s and not s.startswith("#") and not s.startswith("```"):
                summary = s[:200]
                break
        summary = summary or "(no summary)"

    hours = _parse_ttl(ttl)
    payload = {
        "body": body,
        "title": title,
        "summary": summary,
        "ttl_hours": hours,
        "once": once,
    }
    with httpx.Client(base_url=_share_url(), timeout=30, headers=_share_headers()) as client:
        resp = client.post("/share", json=payload)
        if resp.status_code != 200:
            raise click.ClickException(f"{resp.status_code}: {resp.text}")
        data = resp.json()

    url = data["url"]
    expires = data["expires_at"]
    expiry_note = f"expires {expires}" + (" — single-use" if once else "")

    clipboard_text = (
        f"Continue this Claude convo ({expiry_note}):\n"
        f"curl -sSL '{url}'\n"
        f"\n"
        f"About: {summary}\n"
    )

    click.echo("\n" + clipboard_text)
    if not no_copy and _copy(clipboard_text):
        click.echo(f"{click.style('✔ copied to clipboard', fg='green')}", err=True)
    elif not no_copy:
        click.echo(click.style("(pbcopy unavailable — text above is the share)", fg="yellow"), err=True)


@share.command("list")
def list_cmd() -> None:
    """List your active share links."""
    with httpx.Client(base_url=_share_url(), timeout=15, headers=_share_headers()) as client:
        resp = client.get("/share")
        resp.raise_for_status()
        shares = resp.json()

    if not shares:
        click.echo("(no active shares)")
        return
    for s in shares:
        once = " [once]" if s.get("once") else ""
        click.echo(f"{s['token']}{once}  expires {s['expires_at']}  {s.get('title') or ''}")
        click.echo(f"  {s['url']}")


@share.command("revoke")
@click.argument("token")
def revoke_cmd(token: str) -> None:
    """Revoke a share link early."""
    with httpx.Client(base_url=_share_url(), timeout=15, headers=_share_headers()) as client:
        resp = client.delete(f"/share/{token}")
        if resp.status_code == 404:
            raise click.ClickException("Not found or not owned by you")
        resp.raise_for_status()
    click.echo(f"Revoked {token}")

File: internal/cmd/share.go
package cmd

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/Cassandras-Edge/cass/internal/auth"
	"github.com/Cassandras-Edge/cass/internal/share"
)

const shareURLDefault = "https://share.cassandrasedge.com"

func shareCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "share",
		Short: "Share Claude Code conversations via ephemeral URLs",
	}
	cmd.AddCommand(shareCreateCmd())
	cmd.AddCommand(shareListCmd())
	cmd.AddCommand(shareRevokeCmd())
	return cmd
}

func shareBaseURL() string {
	if u := os.Getenv("CASS_SHARE_URL"); u != "" {
		return u
	}
	return shareURLDefault
}

// shareRequest fires a request to the share service with CF Access service-token
// + Bearer headers from ~/.cass/env, then decodes the response.
func shareRequest(method, path string, body, out any) error {
	creds, err := auth.Read()
	if err != nil {
		return fmt.Errorf("not logged in (run: cass login): %w", err)
	}
	var reader io.Reader
	if body != nil {
		buf, err := json.Marshal(body)
		if err != nil {
			return err
		}
		reader = bytes.NewReader(buf)
	}
	req, err := http.NewRequest(method, shareBaseURL()+path, reader)
	if err != nil {
		return err
	}
	req.Header.Set("CF-Access-Client-Id", creds.CFAccessClientID)
	req.Header.Set("CF-Access-Client-Secret", creds.CFAccessClientSecret)
	req.Header.Set("Authorization", "Bearer "+creds.MCPKey)
	if devEmail := os.Getenv("CASS_DEV_EMAIL"); devEmail != "" {
		req.Header.Set("X-Dev-Email", devEmail)
	} else if creds.Email != "" {
		req.Header.Set("X-Dev-Email", creds.Email)
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	resp, err := (&http.Client{Timeout: 30 * time.Second}).Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		buf, _ := io.ReadAll(io.LimitReader(resp.Body, 2048))
		return fmt.Errorf("share %s: %s", resp.Status, string(buf))
	}
	if out == nil {
		return nil
	}
	return json.NewDecoder(resp.Body).Decode(out)
}

func shareCreateCmd() *cobra.Command {
	var ttl, title, summary string
	var once, noCopy bool
	cmd := &cobra.Command{
		Use:   "create [SESSION]",
		Short: "Upload the current (or specified) session as a share link",
		Args:  cobra.MaximumNArgs(1),
		RunE: func(_ *cobra.Command, args []string) error {
			var sessionPath string
			if len(args) == 1 {
				sessionPath = args[0]
			}
			return runShareCreate(sessionPath, ttl, title, summary, once, noCopy)
		},
	}
	cmd.Flags().StringVar(&ttl, "ttl", "24h", "Expiry: e.g. 6h, 24h, 7d")
	cmd.Flags().BoolVar(&once, "once", false, "Self-destruct after first fetch")
	cmd.Flags().StringVar(&title, "title", "", "Optional human title")
	cmd.Flags().StringVar(&summary, "summary", "", "2-3 line summary; defaults to the first user turn")
	cmd.Flags().BoolVar(&noCopy, "no-copy", false, "Skip copying to clipboard")
	return cmd
}

func runShareCreate(sessionArg, ttl, title, summary string, once, noCopy bool) error {
	hours, err := parseTTL(ttl)
	if err != nil {
		return err
	}
	jsonlPath, err := resolveSessionPath(sessionArg)
	if err != nil {
		return err
	}
	fmt.Fprintf(os.Stderr, "Reading %s...\n", jsonlPath)
	f, err := os.Open(jsonlPath)
	if err != nil {
		return err
	}
	body := share.JSONLToMarkdown(f, title)
	f.Close()
	fmt.Fprintf(os.Stderr, "  → %d chars of sanitized markdown\n", len(body))

	if summary == "" {
		for _, line := range strings.Split(body, "\n") {
			s := strings.TrimSpace(line)
			if s == "" || strings.HasPrefix(s, "#") || strings.HasPrefix(s, "```") {
				continue
			}
			if len(s) > 200 {
				s = s[:200]
			}
			summary = s
			break
		}
		if summary == "" {
			summary = "(no summary)"
		}
	}

	payload := map[string]any{
		"body":      body,
		"title":     title,
		"summary":   summary,
		"ttl_hours": hours,
		"once":      once,
	}
	var resp struct {
		URL       string `json:"url"`
		ExpiresAt string `json:"expires_at"`
	}
	if err := shareRequest("POST", "/share", payload, &resp); err != nil {
		return err
	}

	expiryNote := "expires " + resp.ExpiresAt
	if once {
		expiryNote += " — single-use"
	}
	clip := fmt.Sprintf("Continue this Claude convo (%s):\ncurl -sSL '%s'\n\nAbout: %s\n",
		expiryNote, resp.URL, summary)
	fmt.Println()
	fmt.Println(clip)
	if !noCopy && copyToClipboard(clip) {
		fmt.Fprintln(os.Stderr, "✔ copied to clipboard")
	} else if !noCopy {
		fmt.Fprintln(os.Stderr, "(clipboard tool unavailable — text above is the share)")
	}
	return nil
}

func shareListCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "list",
		Short: "List your active share links",
		RunE: func(_ *cobra.Command, _ []string) error {
			var shares []struct {
				Token     string `json:"token"`
				URL       string `json:"url"`
				ExpiresAt string `json:"expires_at"`
				Title     string `json:"title"`
				Once      bool   `json:"once"`
			}
			if err := shareRequest("GET", "/share", nil, &shares); err != nil {
				return err
			}
			if len(shares) == 0 {
				fmt.Println("(no active shares)")
				return nil
			}
			for _, s := range shares {
				once := ""
				if s.Once {
					once = " [once]"
				}
				fmt.Printf("%s%s  expires %s  %s\n", s.Token, once, s.ExpiresAt, s.Title)
				fmt.Printf("  %s\n", s.URL)
			}
			return nil
		},
	}
}

func shareRevokeCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "revoke <token>",
		Short: "Revoke a share link early",
		Args:  cobra.ExactArgs(1),
		RunE: func(_ *cobra.Command, args []string) error {
			if err := shareRequest("DELETE", "/share/"+args[0], nil, nil); err != nil {
				return err
			}
			fmt.Printf("Revoked %s\n", args[0])
			return nil
		},
	}
}

// resolveSessionPath finds the JSONL to share. Precedence: explicit arg,
// $CLAUDE_SESSION_ID lookup under ~/.claude/projects, or newest .jsonl in
// the cwd-hashed project dir.
func resolveSessionPath(arg string) (string, error) {
	if arg != "" {
		expanded, err := expandHome(arg)
		if err != nil {
			return "", err
		}
		if _, err := os.Stat(expanded); err != nil {
			return "", fmt.Errorf("no such file: %s", expanded)
		}
		return expanded, nil
	}
	home, _ := os.UserHomeDir()
	base := filepath.Join(home, ".claude", "projects")
	if _, err := os.Stat(base); err != nil {
		return "", fmt.Errorf("%s missing — Claude Code not set up on this machine", base)
	}
	if sid := os.Getenv("CLAUDE_SESSION_ID"); sid != "" {
		entries, _ := os.ReadDir(base)
		for _, e := range entries {
			if !e.IsDir() {
				continue
			}
			candidate := filepath.Join(base, e.Name(), sid+".jsonl")
			if _, err := os.Stat(candidate); err == nil {
				return candidate, nil
			}
		}
	}
	// Fallback: newest .jsonl in the cwd-hashed dir.
	cwd, _ := os.Getwd()
	mangled := "-" + strings.ReplaceAll(cwd, "/", "-")
	dir := filepath.Join(base, mangled)
	entries, err := os.ReadDir(dir)
	if err != nil {
		return "", fmt.Errorf("could not locate session .jsonl; pass the path explicitly")
	}
	type fileMod struct {
		path string
		mod  time.Time
	}
	var files []fileMod
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".jsonl") {
			continue
		}
		info, err := e.Info()
		if err != nil {
			continue
		}
		files = append(files, fileMod{filepath.Join(dir, e.Name()), info.ModTime()})
	}
	if len(files) == 0 {
		return "", fmt.Errorf("could not locate session .jsonl; pass the path explicitly")
	}
	sort.Slice(files, func(i, j int) bool { return files[i].mod.After(files[j].mod) })
	return files[0].path, nil
}

func expandHome(p string) (string, error) {
	if !strings.HasPrefix(p, "~") {
		return p, nil
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, p[1:]), nil
}

var ttlPattern = regexp.MustCompile(`^(\d+)([hd])$`)

func parseTTL(ttl string) (int, error) {
	m := ttlPattern.FindStringSubmatch(strings.ToLower(strings.TrimSpace(ttl)))
	if m == nil {
		return 0, fmt.Errorf("invalid --ttl %q; use forms like 6h, 24h, 7d", ttl)
	}
	n, _ := strconv.Atoi(m[1])
	if m[2] == "h" {
		return n, nil
	}
	return n * 24, nil
}

func copyToClipboard(s string) bool {
	var c *exec.Cmd
	switch runtime.GOOS {
	case "darwin":
		c = exec.Command("pbcopy")
	case "linux":
		if _, err := exec.LookPath("xclip"); err == nil {
			c = exec.Command("xclip", "-selection", "clipboard")
		} else if _, err := exec.LookPath("wl-copy"); err == nil {
			c = exec.Command("wl-copy")
		}
	}
	if c == nil {
		return false
	}
	c.Stdin = strings.NewReader(s)
	return c.Run() == nil
}
